modules/nlu/src/backend/pipelines/intents/exact_matcher.ts

```diff
@@ -1,18 +1,59 @@
 import * as sdk from 'botpress/sdk'
 import _ from 'lodash'
 
-import { Sequence } from '../../typings'
+import { TrainingSequence, Sequence } from '../../typings'
 import { sanitize } from '../language/sanitizer'
 
 // TODO if we're going to keep this, replace the training set with an inversed index or tree or at anything faster than O(n) at predict time
 // this might be replaced by a knn with tweaked distance func & proper usage at predict time
 export default class ExactMatcher {
-  constructor(private trainingSet: Sequence[]) {}
+  constructor(private trainingSet: TrainingSequence[]) {}
 
   exactMatch(text: string, includedContext: string[]): sdk.NLU.Intent | void {
+    return this._exactMatchFromSequence(text, includedContext, this.trainingSet)
+  }
+
+  exactMatchIgnoringEntity(text: string, includedContext: string[], entity: sdk.NLU.Entity) {
```

Comment:
A few issues here:

1) we don't want to ignore an entity in the user input, we want to ignore a slot in the training sequence (so usage in engine.ts is very weird)
2) we end up generating a new corpus for every incoming messages
3) we're checking for all slots (the requirement is up to 1 slot)

I think a better approach would be to reverse this logic so that:
- `exactMatch()` takes the entire `ds` as input
- the exact matcher, for all "sequence", would have the following logic:

```
matches = []
for ( seq of all_seqs ) {
  if seq not in context, continue

  if seq does not contain slots,
     if seq.canonical === ds.canonical
        matches.push ( confidence: 1, intent )

  else if seq contains 1 slot,
      removable_entities = seq.slot1.entities
      seqNoSlot = seq.canonical.substr(0, slot1.start) + seq.canonical.substr(slot1.end)

         if ( ds.entities.find( e => textWithoutEntity === seqNoSlot ) )
           matches.push ( confidence: 1, intent )
}
```